tests/numerics/ad/test_grid_operators.py

```diff
@@ -0,0 +1,587 @@
+
+    # Check that the projection matrix for the top-dimensional grid is non-zero.
+    # The matrix has eight boundary faces.
+    ind0 = 0
+    ind1 = g_0.num_faces * proj_dim
+    assert np.sum(subdomain_to_boundary[:, ind0:ind1]) == 8 * proj_dim
+    # Check that the projection matrix for the first fracture is non-zero. Since the
+    # fracture touches the boundary on two sides, we expect two non-zero rows.
+    ind0 = ind1
+    ind1 += g_1.num_faces * proj_dim
+    assert np.sum(subdomain_to_boundary[:, ind0:ind1]) == 2 * proj_dim
+    # Check that the projection matrix for the second fracture is non-zero.
+    ind0 = ind1
+    ind1 += g_2.num_faces * proj_dim
+    assert np.sum(subdomain_to_boundary[:, ind0:ind1]) == 2 * proj_dim
+    # The projection matrix for the intersection should be zero.
+    ind0 = ind1
+    assert np.sum(subdomain_to_boundary[:, ind0:]) == 0
+
+    # Make second projection on subset of grids.
+    subdomains = [g_0, g_1]
+    projection = pp.ad.grid_operators.BoundaryProjection(mdg, subdomains, proj_dim)
+    num_faces = proj_dim * (g_0.num_faces + g_1.num_faces)
+    num_cells = proj_dim * sum(
+        [mdg.subdomain_to_boundary_grid(sd).num_cells for sd in subdomains]
+    )
+    # Obtaining sparse matrices from the AD Operators.
+    subdomain_to_boundary = projection.subdomain_to_boundary.parse(mdg)
+    boundary_to_subdomain = projection.boundary_to_subdomain.parse(mdg)
+    # Check sizes.
+    assert subdomain_to_boundary.shape == (num_cells, num_faces)
+    assert boundary_to_subdomain.shape == (num_faces, num_cells)
+
+    # Check that the projection matrix for the top-dimensional grid is non-zero.
+    # Same sizes as above.
+    ind0 = 0
+    ind1 = g_0.num_faces * proj_dim
+    assert np.sum(subdomain_to_boundary[:, ind0:ind1]) == 8 * proj_dim
+    ind0 = ind1
+    ind1 += g_1.num_faces * proj_dim
+    assert np.sum(subdomain_to_boundary[:, ind0:ind1]) == 2 * proj_dim
+
+    # Check that subdomain_to_boundary and boundary_to_subdomain are consistent.
+    assert np.allclose((subdomain_to_boundary - boundary_to_subdomain.T).data, 0)
+
+
+@pytest.mark.integtest
+# Geometry based operators
+def test_trace(mdg: pp.MixedDimensionalGrid):
+    """Test Trace operator.
+
+    Parameters:
+        mdg: Mixed-dimensional grid.
+
+    This test is not ideal. It follows the implementation of Trace relatively closely,
+    but nevertheless provides some coverage, especially if Trace is carelessly changed.
+    The test constructs the expected md trace and inv_trace matrices and compares them
+    to the ones of Trace. Also checks that an error is raised if a non-scalar trace is
+    constructed (not implemented).
+    """
+    # The operator should work on any subset of mdg.subdomains.
+    subdomains = mdg.subdomains(dim=1)
+
+    # Construct expected matrices.
+    traces, inv_traces = list(), list()
+    # No check on this function here.
+    # TODO: A separate unit test might be appropriate.do in this PR?
```

Comment:
Would it make sense to resolve this todo in this PR?

Reply:
I did not understand the comment at all, it gave no sense, so I deleted it.